kratos/containers/csr_matrix.h

```diff
@@ -593,8 +592,45 @@ class CsrMatrix final
     // RightScaling
     // SymmetricScaling
 
-    //TODO
-    //void ApplyDirichlet
+    template<class TVectorType1, class TVectorType2=TVectorType1>
+    void ApplyHomogeneousDirichlet(const TVectorType1& free_dofs_vector, 
+                                    const TDataType DiagonalValue, 
+                                    TVectorType2& rRHS)
+    {
+        KRATOS_ERROR_IF(size1() != free_dofs_vector.size() ) << "ApplyDirichlet: mismatch between row sizes : " << size1()  
+            << " and free_dofs_vector size " << free_dofs_vector.size() << std::endl;
+        KRATOS_ERROR_IF(size2() != free_dofs_vector.size() ) << "ApplyDirichlet: mismatch between col sizes : " << size2()  
+            << " and free_dofs_vector size " << free_dofs_vector.size() << std::endl;
+
+        if(nnz() != 0)
+        {
+            IndexPartition<IndexType>(size1()).for_each( [&](IndexType i){
+
+                //set to zero the relevant row in the RHS
+                rRHS[i] *= free_dofs_vector[i];
+
+                IndexType row_begin = index1_data()[i];
+                IndexType row_end   = index1_data()[i+1];
+                if(std::abs(free_dofs_vector[i]-1.0) < 1e-10) //row corresponding to free dofs NOTE that we check if it is approx zero
```

Comment:
1e-10 seems rather large no?
Did you choose this from experience?

Reply:
the values should be either 1 or 0

so an error of 1e-10 is 10 orders of magnitude smaller (and it needs only to take into account smaller communication-related glitches)

Reply:
ah ok I didn't understand that those are used for this purpose